infrastructure/containerd/image_service_integration_test.go

```diff
@@ -0,0 +1,291 @@
+package containerd_test
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+
+	ctr "github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/snapshots"
+	. "github.com/onsi/gomega"
+
+	"github.com/weaveworks/flintlock/core/models"
+	"github.com/weaveworks/flintlock/core/ports"
+	"github.com/weaveworks/flintlock/infrastructure/containerd"
+)
+
+const (
+	testImageVolume    = "docker.io/library/alpine:3.14.1"
+	testImageKernel    = "docker.io/linuxkit/kernel:5.4.129"
+	testSnapshotter    = "devmapper"
+	testOwnerNamespace = "int_ns"
+	testOwnerUsageID   = "vol1"
+	testOwnerName      = "imageservice-get-test"
+	testContainerdNS   = "flintlock_test_ctr"
+)
+
+// What this test does?
+//
+// == Preface
+//
+// Prepare ImageService and a Namespace context. All images and snapshots will
+// live under the same namespace.
+//
+// We do some prudent cleanup steps in case the test was aborted. Because it's
+// an integration test with an external service, if we don't clean up after
+// ourselves, the next test run will most likely fail. In theory, releasing the
+// leases should be enough, but that's heavily depend on how containerd tells
+// us how it works, so it's easier to delete everything we intended to create.
+//
+// == Chapter I: Pull and mount
+//
+// Using ImageService, we tries to Pull an image.
+//
+// We don't have separate Mount function (yet), we try to PullAndMount, with
+// this, we can be sure ImageService can pull and image from a repository and
+// can mount it.
+//
+// To avoid false positive result, we try to PullAndMount an image, we know
+// it's definitely not there, and we expect it to be failed.  If it does not
+// fail, something fishy is swimming under the hood.
+//
+// After PullAndMount, the real image should be mounted, but the fake one
+// shouldn't be. At the end we should have only one image available locally.
+//
+// == Chapter II: Snapshots
+//
+// As we already pull and mounted an image, we expect one snapshot to be there
+// with the name constructed from the VM namespace, the VM name, and the volume
+// usage type.
+// At the same time, we expect one lease to be there with similar naming
+// structure except the volume usage type we have one lease for all resources
+// in containerd.
+//
```

Comment:
can we clarify that this is not a separate test stage in itself. we are not checking any new code here, it is part of the PullAndMount checks

Reply:
actually it could be moved after the kernel Pull and then we can check that all the snapshots are what we expect like in a loop or smth?

Reply:
The logic is there to mimic the usage. (we actually don't even use `Pull` only `PullAndMount`, it's just an extra test). So `Chapter I` is a full workflow we are using on CreateVM on one volume. The Kernel does the same, but it does some bits different, but overall the same, so that test was added to make sure the PullAndMouont part is working on kernel images too, but the rest handled the same.

Reply:
> we actually don't even use Pull

lol! why are we keeping it? for a rainy day?

Reply:
maybe :rofl: